### ISSUE TITLE
[Bug]: cert-manager.io/cluster-issuer annotation does not work for AWSPCAClusterIssuer

### ISSUE THREAD

**Issue by brsolomon-deloitte** [cert-manager/aws-privateca-issuer#252]
### Describe the expected outcome

The `cert-manager.io/cluster-issuer` annotation for `Ingress` should be allowed to point to a `AWSPCAClusterIssuer`.

### Describe the actual outcome

To refer to a `AWSPCAClusterIssuer`, cert-manager _only_ recognizes `cert-manager.io/issuer` and _not_ `cert-manager.io/cluster-issuer`.

This is counter-intuitive given that `cluster-issuer` is supposed to point to a non-namespaced resource.

https://cert-manager.io/docs/usage/ingress/#supported-annotations

### Steps to reproduce

1. Deploy cert-manager and aws-privateca-issuer using Helm
2. Create a `AWSPCAClusterIssuer`
3. Create a test app with an `Ingress` and annotate it with `cert-manager.io/cluster-issuer`

The result is that nothing happens at all - the annotation is effectively ignored.

### Version

1.2.4

### Have you tried the following?

- [X] Check the [Troubleshooting section](../#troubleshooting)
- [X] Search open [issues](https://github.com/cert-manager/aws-privateca-issuer/issues)

### Category

Other

### Severity

Severity 3

**Comment by brsolomon-deloitte** [comment]
If this is somehow the intended behavior - then it doesn't make much sense.

`cert-manager.io/issuer` should point to an `AWSPCAIssuer`

`cert-manager.io/cluster-issuer` should point to an `AWSPCAClusterIssuer`

This does not seem to be formally documented nor tested.

**Comment by KyleBS** [comment]
Thank you for raising this issue with the AWS Private CA Issuer plugin. We will review your submission and respond back to you here as soon as possible.

**Comment by Hamidhasan** [comment]
I was able to test this offline, and I can confirm that even with just the `cluster-issuer` annotation (i.e. not specifying the kind or group), we're unable to issue the certificate. The regular issuer, when using the kind and group annotations, is able to issue it, but the cluster issuer is not. Here's the error messages I get (it seems like cert-manager attempts to use a few different issuer types but they don't include ours, and thus the cert issuance fails):

```
  Normal  WaitingForApproval  53s   cert-manager-certificaterequests-issuer-ca          Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  53s   cert-manager-certificaterequests-issuer-vault       Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  53s   cert-manager-certificaterequests-issuer-venafi      Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  53s   cert-manager-certificaterequests-issuer-selfsigned  Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  53s   cert-manager-certificaterequests-issuer-acme        Not signing CertificateRequest until it is Approved
  Normal  IssuerNotFound      53s   cert-manager-certificaterequests-issuer-selfsigned  Referenced "ClusterIssuer" not found: clusterissuer.cert-manager.io "pca-cluster-issuer-rsa" not found
  Normal  IssuerNotFound      53s   cert-manager-certificaterequests-issuer-acme        Referenced "ClusterIssuer" not found: clusterissuer.cert-manager.io "pca-cluster-issuer-rsa" not found
  Normal  IssuerNotFound      53s   cert-manager-certificaterequests-issuer-ca          Referenced "ClusterIssuer" not found: clusterissuer.cert-manager.io "pca-cluster-issuer-rsa" not found
  Normal  IssuerNotFound      53s   cert-manager-certificaterequests-issuer-vault       Referenced "ClusterIssuer" not found: clusterissuer.cert-manager.io "pca-cluster-issuer-rsa" not found
  Normal  IssuerNotFound      53s   cert-manager-certificaterequests-issuer-venafi      Referenced "ClusterIssuer" not found: clusterissuer.cert-manager.io "pca-cluster-issuer-rsa" not found
  Normal  cert-manager.io     53s   cert-manager-certificaterequests-approver           Certificate request has been approved by cert-manager.io
```

For reference, here is the yaml file I applied in order to get those messages.

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    # add an annotation indicating the issuer to use.
    cert-manager.io/cluster-issuer: pca-cluster-issuer-rsa
  name: my-ingress-cluster-issuer
  namespace: aws-privateca-issuer
spec:
  rules:
  - host: example.com
    http:
      paths:
      - pathType: Prefix
        path: /
        backend:
          service:
            name: myservice
            port:
              number: 80
  tls: # < placing a host in the TLS config will determine what ends up in the cert's subjectAltNames
  - hosts:
    - example.com
    secretName: myingress-issuer-cert # < cert-manager will store the created certificate in this secret.
```

There is a biweekly sync with cert-manager, we'll raise this issue with them and see what they say. There may need to be some support on cert-manager's side to make this work correctly.

**Comment by brsolomon-deloitte** [comment]
@Hamidhasan 

> Can you post the Ingress resource you’ve tested this with. ... For reference, here is the yaml file I applied in order to get those messages.

Seems like you've been able to reproduce exactly what we are seeing, e.g. that a `kind: Ingress` definition will fail to issue a cert when using `cert-manager.io/cluster-issuer`, but will succeed with `cert-manager.io/issuer`, even though the issuer in question is a non-namespaced cluster-wide `AWSPCAClusterIssuer`.

**Comment by Hamidhasan** [comment]
Leaving some more notes in here, plan to present to cert-manager weekly meeting:

This is related to the [securing ingresses](https://cert-manager.io/docs/usage/ingress/) feature of cert-manager.

For our aws-privateca-issuer, only `cert-manager.io/issuer` annotation will work, and only if it is combined with the `cert-manager.io/issuer-kind` and `cert-manager.io/issuer-group` annotations. The other 3 combinations fail:
1. Just `issuer`, fails with the error messages above
2. Just `cluster-issuer`, fails with the error messages above
3. `cluster-issuer` with `issuer-kind` and `issuer-group`, fails due to the check in cert-manager [ingress-shim code](https://github.com/cert-manager/cert-manager/blob/706ad574b9bb7669ae94d293d4136cc37cb7d09f/pkg/controller/certificate-shim/sync.go#L705-L715). 

Basically, we are not able to get the ingress to automatically obtain a cert in the case of a cluster-issuer.

**Comment by Hamidhasan** [comment]
One question to you @brsolomon-deloitte is: while this is not ideal, does the potential workaround of specifying the cluster issuer in the `issuer` annotation actually work for you? 

Put another way: if this was fixed, does it unblock/enable any more functionality for your use cases, or can you use the workaround to do everything you need?

**Comment by brsolomon-deloitte** [comment]
> while this is not ideal, does the potential workaround of specifying the cluster issuer in the issuer annotation actually work for you?

Yes, we realize that using `issuer` to point to a `AWSPCAClusterIssuer` does _work_, when combined with `issuer-kind`. However, from our experience, it seems buggy and confusing from a usability perspective, and doesn't seem to match with the semantics/definitions of `issuer` versus `cluster-issuer`, since `AWSPCAClusterIssuer` arguably best fits with `cluster-issuer`. At the very least, it seems like this behavior should be documented, either by `aws-privateca-issuer` or `cert-manager`.

Either way, thanks for laying out the details and giving attention to this issue.

**Comment by divyansh-gupta** [comment]
Thanks for creating this issue @brsolomon-deloitte and thank you @irbekrm for the help in fixing the documentation! I am closing this issue now.
